### ISSUE TITLE
event latency unit

### ISSUE THREAD

**Issue by I-am-not-a-number** [anarchocurious/flic#1]
Hi there,
the example code suggests that `latency_ms` will show milliseconds, but it seems to use seconds.

```
...
2016-12-11 18:53:54 +0100 [xx.xx.xx.xx] Handling button_up latency is 0 ms
2016-12-11 18:54:02 +0100 [xx.xx.xx.xx] Handling button_single_click latency is 8 ms
2016-12-11 18:54:02 +0100 [xx.xx.xx.xx] Handling button_down latency is 0 ms
...
```

Maybe I did something wrong, as "fast" single clicks are only shown as

```
2016-12-11 18:54:02 +0100 [xx.xx.xx.xx] Handling button_down latency is 0 ms
2016-12-11 18:54:02 +0100 [xx.xx.xx.xx] Handling button_up latency is 0 ms
```

and the next (a bit longer, but not yet "hold" ;-) click will return two single clicks

```
2016-12-11 18:56:47 +0100 [xx.xx.xx.xx] Handling button_single_click latency is 164 ms
2016-12-11 18:56:47 +0100 [xx.xx.xx.xx] Handling button_down latency is 0 ms
2016-12-11 18:56:48 +0100 [xx.xx.xx.xx] Handling button_up latency is 0 ms
2016-12-11 18:56:48 +0100 [xx.xx.xx.xx] Handling button_single_click latency is 0 ms
```

Any idea?

Thanks and cheers,
Markus

**Comment by Emill** [comment]
If it just forwards the raw value from the flicd, then it is seconds and not milliseconds. The value should only be non-zero if the button was in the disconnected state when it was pressed.

**Comment by alecdotninja** [comment]
@Emill Is correct; the current implementation just forwards [the `time_diff` field from the button event messages](https://github.com/50ButtonsEach/fliclib-linux-hci/blob/master/ProtocolDocumentation.md#evtbuttonevent).

I'm not sure where I got the idea that this field contains the number of milliseconds. If I have time later this week, I will go through and refactor the variable names / update the documentation accordingly.

**Comment by I-am-not-a-number** [comment]
Thanks guys!!
Is there any way I can get the ms or even better the `single_click` event, without waiting for the next time someone pushes the flic?

**Comment by Emill** [comment]
How do you mean? If it's connected, you will get the event immediately.
If it's disconnected, it'll try to connect to your host.

**Comment by I-am-not-a-number** [comment]
If I click the flic "too fast" I just get the `button_down` and `button_up` event, but not the `single_click` event. So I'm searching either for a more precise delay info (in ms) to make my own checks or a way receive the `single_click` in time. (Preferably using the `Flic::SimpleClient`:sleeping:)

**Comment by Emill** [comment]
Do you mean you get a double click event when you press too many times fast?

**Comment by I-am-not-a-number** [comment]
No, it's just about a single click. If I release the button too fast, it does not send a single click event.

Pressing and releasing the Flic to fast gives me:
```
2016-12-12T21:37:01.562+01:00 [xx.xx.xx.xx] Handling button_down latency is 0 ms
2016-12-12T21:37:01.861+01:00 [xx.xx.xx.xx] Handling button_up latency is 0 ms
```

Then I wait some seconds and do the same thing again
```
2016-12-12T21:37:17.054+01:00 [xx.xx.xx.xx] Handling button_single_click latency is 15 ms <-- this is the old click, I wanted to know about 15 seconds earlier ;-)
2016-12-12T21:37:17.059+01:00 [xx.xx.xx.xx] Handling button_down latency is 0 ms
2016-12-12T21:37:17.402+01:00 [xx.xx.xx.xx] Handling button_up latency is 0 ms
```

This time I push the Flic just a bit longer and get two events ...
```
2016-12-12T21:37:32.920+01:00 [xx.xx.xx.xx] Handling button_single_click latency is 15 ms <-- old click
2016-12-12T21:37:32.924+01:00 [xx.xx.xx.xx] Handling button_down latency is 0 ms
2016-12-12T21:37:33.813+01:00 [xx.xx.xx.xx] Handling button_up latency is 0 ms
2016-12-12T21:37:33.820+01:00 [xx.xx.xx.xx] Handling button_single_click latency is 0 ms <-- current click
```

I think this is unrelated to the ruby flic library, so if I should read some documentation or post my issues somewhere else, just let me know. :smile: 

Thanks!

**Comment by alecdotninja** [comment]
As I understand it, the latency is the amount of time between the event being dispatched from `flicd` and the button being pressed. It is usually zero (especially in a `Flic::Client::ConnectionChannel` whose `latency_mode` is `:low`), but may be longer if the button is not connected to the host at the time it is pressed. ASFAIK, buttons will hold events essentially indefinitely until they connect to the host again (can you confirm that @Emill?).

All that is to say, the underlying latency is unrelated to how a sequence of physical actions on a button is mapped to events like `:button_up`, `:button_down`, `:button_single_click`, etc. Sadly, ASFAIK, these mappings are a non-configurable aspect of `flicd` (again, perhaps @Emill can confirm this).

**Comment by alecdotninja** [comment]
@I-am-not-a-number Looking at the output from your latest comment, it would seem that `flicd` is delaying the `:button_single_click` event 15 seconds (and indeed reporting that it is doing exactly that!). If the Ruby client was the source of the delay, I would expect it to report a latency less than the delay (probably zero).

I agree that it looks like this may be an issue with `flicd`.

**Comment by alecdotninja** [comment]
@I-am-not-a-number Sorry to spam you, but I just thought of something. Have you tried passing `:low` as the first argument to `Flic::SimpleClient#listen`? This may coincidentally solve your problem by setting the `latency_mode` of the underlying `Flic::Client::ConnectionChannel` to `:low`.

**Comment by Emill** [comment]
Could you try one of the official clients (either one in https://github.com/50ButtonsEach/fliclib-linux-hci/tree/master/clientlib or https://github.com/50ButtonsEach/fliclib-linux-hci/tree/master/simpleclient)?

All events are created in the button and are stored indefinitely until delivered (but has a circle-buffer to remove old events on overflow). The time_diff is also created inside the button and is the time from when the event clicked until it was sent over the air. Only if the flic was disconnected the time_diff may be non-zero.

**Comment by I-am-not-a-number** [comment]
@anarchocurious 
 I've been using `:low` mode with `Flic::SimpleClient` already. Is there any other setting that might help? 

@Emill 
`fliclib-linux-hci/simpleclient/simpleclient` works fine
```
Button up/down: 2, ButtonDown, not queued, 0 seconds ago
Button up/down: 2, ButtonUp, not queued, 0 seconds ago
Button click/hold: 2, ButtonClick, not queued, 0 seconds ago
Button single/double click: 2, ButtonSingleClick, not queued, 0 seconds ago
Button single/double click/hold: 2, ButtonSingleClick, not queued, 0 seconds ago
```
I don't know how to add timestamps here, but there's a split second delay before the last two lines appear (the delay used to detect double clicks I guess), but I get all the events I expect.

So to me it seems `Flic::SimpleClient` disconnects too fast. Is this a power saving thing? Or could a Flic stay connected all the time?

**Comment by Emill** [comment]
I noticed a documentation error in the ProtocolDocumentation.md: https://github.com/50ButtonsEach/fliclib-linux-hci/commit/4cd30d0e612bc1c1ec699be6f4eeb0b08b63fbc9. The ruby client sets value 512 which will be treated as 0. Therefore it will disconnect immediately after all button events have been delivered. The correct value to use to stay connected is 511. Please replace all 512 with 511 here: https://github.com/anarchocurious/flic/blob/master/lib/flic/protocol/primitives/disconnect_time.rb.

Setting the "latency mode" has only impact of the connection parameters used, which means how often the button and the host will exchange packets over the air.

**Comment by alecdotninja** [comment]
Excellent catch, @Emill! It seems that I may have been having a different manifestation of @I-am-not-a-number's problem all along. In my case, however, the events were always delivered within a few seconds. Since I've updated the value, my home automation script registers their presses virtually instantaneously. Thank you!

I recall being suspicious of that value when I was first implementing the protocol, but I didn't give too much thought to it since the field is a `uint32`. If you don't mind, I'm curious, why is that field a `uint32` in the protocol if it is clamped at `uint8`?

@I-am-not-a-number I've pushed [a new release](https://rubygems.org/gems/flic/versions/0.0.6) that incorporates this fix. Hopefully this fixes your issue.
